### ISSUE TITLE
Unable to resolve classname: Appendable (clojure.data.json)

### ISSUE THREAD

**Issue by onetom** [babashka/babashka#919]
**version**

babashka v0.4.6

**platform**

macOS

**problem**

Can't load `clojure.data.json`.

**repro**

```
$ echo '{:deps {org.clojure/data.json {:mvn/version "2.3.1"}}}' > bb.edn
$ bb -e '(require (quote [clojure.data.json :as j])) (j/read-str "{\"a\":1}")'
----- Error --------------------------------------------------------------------
Type:     clojure.lang.ExceptionInfo
Message:  Unable to resolve classname: Appendable
Phase:    analysis
```

**expected behavior**

```
$ bb --clojure -Sdeps '{:deps {org.clojure/data.json {:mvn/version "2.3.1"}}}' -M -e '(require (quote [clojure.data.json :as j])) (j/read-str "{\"a\":1}")'
{"a" 1}
```

I've tested with data.json 1.0.0 and bb 0.4.5 too; same result.

**Comment by borkdude** [comment]
@onetom We can add the interface `Appendable` to bb. But even then I'm not sure if clojure.data.json will work from source with bb, but worth a try. I'll try later this week.

**Comment by onetom** [comment]
Maybe it's worth mentioning why have I encountered this issue. While it's great to have the most performant JSON codec - Cheshire - shipped with babashka, in Clojure projects, it's a source of dependency conflicts. Because of that, we standardized on using `clojure.data.json` instead, because of its zero-dependency nature.

It's often desirable to run the configuration related code from such Clojure projects via Babashka, to quickly compute some derived configuration, which follows some naming conventions. We would use such values to integrate with other command line tools. This way we can avoid duplicating naming conventions in `Makefile`s for example.

Now, that I've explained it, I've realized that I might be able to apply some reader conditionals to solve this issue. We already rely on a common abstraction layer, in form of `xxx.common/json-{encode,decode}` functions, to unify behaviour between CLJ & CLJS:

```clojure
(defn json-encode
  "JSON encode `data` and pretty-print it.
   Map keys are encoded as stringified symbols, to preserve the namespace
   of the keys, but avoid extra noise of the `:` prefix.
   Map values are still serialized with `str`, so a keyword map value will still
   have the colon prefix, to encode its original type."
  [data]
  #?(:clj  (with-out-str
             (json/pprint data
                          :escape-slash false
                          :key-fn kw->sym->str))

     :cljs (let [json-pretty-stringify #(gcall :JSON.stringify % nil 2)]
             (-> data
                 stringify-namespaced-keys
                 clj->js
                 json-pretty-stringify))))

(defn json-decode
  "Blank strings are decoded as nil.
   Keys are converted into potentially namespaced keywords.
   To keep keys as strings, use `(json-decode json-str :key-fn str)`."
  [json-str & args]
  (if (str/blank? json-str)
    nil
    #?(:clj  (apply json/read-str json-str :key-fn keyword args)

       :cljs (let [{:keys [key-fn]} (apply hash-map args)]
               (-> (gcall :JSON.parse json-str)
                   (js->clj :keywordize-keys
                            (or (nil? key-fn) (= key-fn keyword))))))))
```

I will try to sprinkle some `:bb: branch into those and see how far I get with that.

**Comment by borkdude** [comment]
I've made this fork of data.json (with very minor tweaks) https://github.com/babashka/data.json work together with babashka on master. I'm not sure how performant it is, as the JSON library is fully interpreted from source, so it needs some testing.

**Comment by borkdude** [comment]
I've tested 10kb of JSON, it parses in 48ms on my machine vs 8ms in Clojure. Not bad at all, considering that all source is interpreted.

**Comment by borkdude** [comment]
100kb of JSON: 350ms in bb vs 24ms in Clojure. I'm sure there are some optimizations to be made, especially around interop, but it's workable if the JSON payloads aren't too big I would say.

**Comment by borkdude** [comment]
More details.

Parsing 100kb of JSON

bb
load time of library: ~20ms
parse time: ~360ms
total time including cold boot: ~420ms

clojure
load time of library: ~350ms
parse time: ~24ms
total time including cold boot: ~1379ms

Note that bb also has cheshire built-in and parsing would be much faster. This is code interpreted from source, which makes it a lot slower. The interop code can be optimized a lot more.

**Comment by onetom** [comment]
What am I missing?
```
$ cat bb.edn  
{:deps {org.clojure/data.json {:git/url "https://github.com/babashka/data.json.git"
                               :sha "7d4ae6a2923965bf2e83962f5268f69eaa4a9a57"}}}

$ bb -e '(require (quote [clojure.data.json :as j])) (j/read-str "{\"a\":1}")'
Cloning: https://github.com/babashka/data.json.git
Checking out: https://github.com/babashka/data.json.git at 7d4ae6a2923965bf2e83962f5268f69eaa4a9a57
----- Error --------------------------------------------------------------------
Type:     clojure.lang.ExceptionInfo
Message:  Unable to resolve classname: Appendable
Phase:    analysis
```

**Comment by borkdude** [comment]
You need the babashka binary from master.

**Comment by onetom** [comment]
My main reason for wishing `clojure.data.json` support in Babashka is allowing the execution of unmodified Clojure code. Typically code, which needs to deal with very little amount of JSON, like reading some OAuth token from the JSON config file of some other program, or performing a few requests against HTTP APIs. So either I don't read a lot of JSON, or network calls would dominate the total runtime of a script.

In other words, saving 1s on Clojure startup time would allow reading ~1MB of JSON instead, which is plenty.

Just as a random, concrete example, https://github.com/XeroAPI/Xero-OpenAPI/blob/master/xero_accounting.yaml in JSON format is 625Kb.

**Comment by borkdude** [comment]
@onetom You can grab a babashka binary either via the commit links to CircleCI or in the babashka-circleci-builds channel on Clojurians slack, to try it out.

**Comment by onetom** [comment]
I just remembered, you already taught me to get any build from that `#babashka-circleci-builds` channel!

Here is a more concrete result:
```
$ ls -lah xero_accounting.json 
-rw-r--r--  1 onetom  staff   625K Jul 10 22:23 xero_accounting.json

$ time ~/Downloads/bb -e '(time (require (quote [clojure.data.json :as j]))) (time (j/read-str (time (slurp "xero_accounting.json")))) :ok'       
"Elapsed time: 21.020285 msecs"     # require
"Elapsed time: 5.608491 msecs"      # slurp
"Elapsed time: 5239.130007 msecs"  # json read-str
:ok
~/Downloads/bb -e   5.19s user 0.11s system 99% cpu 5.317 total
```

**Comment by borkdude** [comment]
So 1mb of JSON takes 5 seconds... yikes :)

**Comment by onetom** [comment]
As opposed to Cheshire's 17ms:

```
$ time ~/Downloads/bb -e '(->> (time (cheshire.core/parse-string (time (slurp "xero_accounting.json")) true)) :paths keys (run! println))'
"Elapsed time: 5.790139 msecs"
"Elapsed time: 16.959945 msecs"
:/Invoices/{InvoiceID}/Attachments/{AttachmentID}
:/Items/{ItemID}
:/Currencies
...
```

Not great, indeed. :/

**Comment by borkdude** [comment]
I think most of this can be explained by the non-optimized interop. For every character `(.read ...)` does a lot more work than probably needed. With more effort I can make this faster, but it will never be as fast as cheshire ;).

**Comment by onetom** [comment]
Given the hack needed to `clojure.data.json` and the resulting abysmal performance, I would not mind to close this issue for now.
It would be great to keep the `Appendable` support in Babashka though, because for small JSON files, this solution is still a feasible choice.

**Comment by borkdude** [comment]
Yes, I will add tests for clojure.data.json in CI.

After fixing another tiny bug, now all its tests are passing, so this exercise was definitely worth it.

```
$ clojure -Sdeps '{:deps {org.clojure/clojure {:mvn/version "1.11.0-alpha1"}}}' -M:babashka/dev -cp src/main/clojure:src/test/clojure -e "(require 'clojure.test 'clojure.data.json-test) (clojure.test/run-tests 'clojure.data.json-test)"
Testing clojure.data.json-test
Ran 66 tests containing 100 assertions.
0 failures, 0 errors.
{:test 66, :pass 100, :fail 0, :error 0, :type :summary}

$ clojure -Sdeps '{:deps {org.clojure/clojure {:mvn/version "1.11.0-alpha1"}}}' -M:babashka/dev -cp src/main/clojure:src/test/clojure -e "(require 'clojure.test 'clojure.data.json-test-suite-test) (clojure.test/run-tests 'clojure.data.json-test-suite-test)"
Testing clojure.data.json-test-suite-test
Ran 128 tests containing 128 assertions.
0 failures, 0 errors.
{:test 128, :pass 128, :fail 0, :error 0, :type :summary}
```

**Comment by onetom** [comment]
For the record, in my benchmarks above, `~/Downloads/bb` came from:
https://20718-201467090-gh.circle-artifacts.com/0/release/babashka-0.4.7-SNAPSHOT-macos-amd64.tar.gz

and I used this `bb.edn`:
```edn
{:paths ["."]
 :deps  {org.clojure/data.json {:git/url "https://github.com/babashka/data.json.git"
                                :sha     "7d4ae6a2923965bf2e83962f5268f69eaa4a9a57"}}}
```

**Comment by borkdude** [comment]
An alternative would be to add clojure.data.json to bb itself, but perhaps its better to invent a community library that dispatches on the concretely provided JSON library while programming against a more abstract interface. This way library users can decide themselves which JSON library to use under the hoods. I've considered this more often. The only issue is that this library should be adopted by everyone ;).

**Comment by onetom** [comment]
Better don't bloat Babashka with `clojure.data.json`.
Instead, we can evolve a recipe on a wiki page, how to have a unified JSON encode/decode function, which works across most runtimes, for the most common case.

Something like:
```clojure
(defn json-decode [json-str & args]
  (let [{:keys [key-fn]} (apply hash-map args)
        keywordize? (or (nil? key-fn) (= key-fn keyword))]
    #?(:bb   (cheshire.core/parse-string json-str keywordize?)
       :clj  (apply json/read-str json-str :key-fn (or key-fn keyword) args)
       :cljs (-> (js/JSON.parse json-str)
                 (js->clj :keywordize-keys keywordize?)))))
```
Then, this can evolve independently from Babashka.

**Comment by borkdude** [comment]
@onetom FWIW, I'm adding the `clojure.tools.logging` abstract interface to bb because of compatibility with libraries in the ecosystem, not for concrete scripts, because many libraries depend on this. That would also be my reasoning for including the `clojure.data.json` shim implemented using cheshire.
